# Programming 5 - Python Programming

## Data Types & Operators

In this lesson, we'll explore **data types** (Numbers, Strings, Booleans) and learn about **operators** (Arithmetic, Comparison, Logical). These are the building blocks of any Python program. By understanding how each data type behaves with different operators, you’ll gain a solid foundation for more complex concepts like data structures, conditionals, and functions.

> **Note**: For a more in-depth discussion on Python’s data types and operators, see  
> - Sanner, M. F. (1999). *Python: A Programming Language for Software Integration and Development.* Journal of Molecular Graphics and Modelling, 17(1), 57-61.  
> - Lutz, M. (2013). *Learning Python* (5th ed.). O’Reilly Media, Inc.  
> - Downey, A. B. (2015). *Think Python: How to Think Like a Computer Scientist.* O’Reilly Media, Inc.  

These resources (all available via Google Scholar) discuss Python’s design philosophy, emphasize clarity over complexity, and provide more detailed explanations and examples.

---


## 1. Overview of Key Concepts

### 1.1 Data Types

1. **Numbers**  
   - **Integers** (e.g. `10`, `-3`) — used for whole numbers (positive, negative, or zero).  
   - **Floats** (e.g. `3.14`, `-0.5`) — used for decimal or fractional values.

2. **Strings**  
   - Represent text enclosed within quotes (`"Hello World"` or `'Hi'`).  
   - Strings are immutable, meaning you can’t change individual characters within them once they’re created (you can only create new strings).

3. **Booleans**  
   - Represent logical values: `True` or `False`.  
   - Often used in conditional statements (`if`, `while`) to control the flow of a program.


---
### 1.2 Operators

1. **Arithmetic Operators**  
   - `+` (Addition)  
   - `-` (Subtraction)  
   - `*` (Multiplication)  
   - `/` (Division)  
   - `//` (Floor Division) — divides and returns the integer part (floored result).  
   - `**` (Exponent) — raises one number to the power of another.  
   - `%` (Modulus) — returns the remainder of a division.

2. **Comparison Operators** (Result is always Boolean: `True` or `False`)  
   - `==` (Equal to)  
   - `!=` (Not equal to)  
   - `>`  (Greater than)  
   - `<`  (Less than)  
   - `>=` (Greater than or equal to)  
   - `<=` (Less than or equal to)

3. **Logical Operators**  
   - `and` — returns `True` if **both** operands are `True`.  
   - `or`  — returns `True` if **at least one** operand is `True`.  
   - `not` — inverts a Boolean value: `True` → `False`; `False` → `True`.

---

In [6]:
# Suppose a customer buys 3 apples at $2 each and 2 oranges at $1.5 each
apples = 3
oranges = 2

In [3]:
# Arithmetic operators to calculate total cost
cost_apples = apples * 2        # 3 * 2 = 6
cost_oranges = oranges * 1.5    # 2 * 1.5 = 3.0


In [4]:
total_cost = cost_apples + cost_oranges  # 6 + 3.0 = 9.0

In [5]:
print("Total Cost of Fruits:", total_cost)

Total Cost of Fruits: 9.0


**Detailed Explanation**

1. **Variable Declarations**  
   - `apples = 3` means we're storing the integer value `3` in the variable `apples`.  
   - `oranges = 2` stores the integer `2`.

2. **Multiplication Operator `*`**  
   - `cost_apples = apples * 2` uses the multiplication operator. Since `apples` is `3`, `cost_apples` becomes `6`.  
   - `cost_oranges = oranges * 1.5` uses a float `1.5`. This results in `3.0`.

3. **Addition Operator `+`**  
   - `total_cost = cost_apples + cost_oranges` adds up `6` and `3.0`, giving us `9.0`.

4. **Print Statement**  
   - `print("Total Cost of Fruits:", total_cost)` outputs **Total Cost of Fruits: 9.0**.

This scenario illustrates how arithmetic operators work with both **integers** and **floats**, automatically producing a **float** result when mixing the two.


---

In [7]:
# Student's score out of 100
score = 85

# Compare the score with a threshold
passed = score >= 75

if passed:
    print("Student passed the exam.")
else:
    print("Student failed the exam.")

Student passed the exam.


**Detailed Explanation**

1. **Comparison Operator `>=`**  
   - `score >= 75` checks if `85` is greater than or equal to `75`. This returns `True`.

2. **Boolean Variable**  
   - `passed` stores the result of that comparison (`True`).

3. **If Statement**  
   - `if passed:` is the same as `if passed == True:`, so the code inside the `if` block runs if `passed` is `True`.

4. **Output**  
   - Because `passed` is `True`, the program prints **"Student passed the exam."**

This demonstrates how **comparison operators** produce a **Boolean** value that can be used to control **program flow** via conditional statements.


---

In [8]:
# Gather user first and last name
first_name = "Alice"
last_name = "Wonderland"

# Concatenate using '+' operator
full_name = first_name + " " + last_name

print("Hello, " + full_name + "!")


Hello, Alice Wonderland!


**Detailed Explanation**

**String Variables:**  
`first_name = "Alice"` and `last_name = "Wonderland"` assign string values to the variables.

**Concatenation with `+`:**  
`full_name = first_name + " " + last_name` joins `"Alice"`, `" "`, and `"Wonderland"` into `"Alice Wonderland"`.

**Print Statement:**  
`print("Hello, " + full_name + "!")` uses the `+` operator to form the final greeting string `"Hello, Alice Wonderland!"`.

When using the `+` operator on strings, Python **concatenates** them, forming one continuous string. This is one way to combine text dynamically.


---

In [9]:
# Roles in a system
is_admin = False
is_editor = True

# Logical operators
if is_admin or is_editor:
    print("You have access to edit content.")
else:
    print("Access denied.")


You have access to edit content.


**Detailed Explanation**

**Boolean Variables:**  
`is_admin = False`  
`is_editor = True`  

**Logical Operator `or`:**  
The expression `is_admin or is_editor` checks if *at least one* is `True`.  
Since `is_editor` is `True`, the entire expression evaluates to `True`.

**Conditional Check:**  
Because the condition is `True`, the program prints `"You have access to edit content."`

**Note on other logical operators:**  
- `and` requires **both** operands to be `True`.  
- `not` flips a Boolean value from `True` to `False` or vice versa.

This is essential in **access control** scenarios, where user permissions rely on Boolean checks.


---

In [10]:
# Input strings representing numbers
food_cost_str = "15.50"
drink_cost_str = "2.45"
tip_percentage_str = "10"

# Convert strings to float or int
food_cost = float(food_cost_str)
drink_cost = float(drink_cost_str)
tip_percentage = int(tip_percentage_str)

# Calculate total
sub_total = food_cost + drink_cost
tip_amount = sub_total * (tip_percentage / 100)
grand_total = sub_total + tip_amount

print("Subtotal:", sub_total)
print("Tip Amount:", tip_amount)
print("Grand Total:", grand_total)


Subtotal: 17.95
Tip Amount: 1.795
Grand Total: 19.744999999999997


**Detailed Explanation**

**String Inputs**  
- Variables like `food_cost_str = "15.50"` look like numbers but are actually **strings**.

**Type Conversion**  
- `float(food_cost_str)` turns the **string** `"15.50"` into a **float** `15.50`.  
- `int(tip_percentage_str)` turns the **string** `"10"` into an **integer** `10`.

**Arithmetic Operations**  
- `sub_total = food_cost + drink_cost` → **15.50 + 2.45 = 17.95**.  
- `tip_amount = sub_total * (tip_percentage / 100)` → **17.95 * (10 / 100) = 1.795**.  
- `grand_total = sub_total + tip_amount` → **17.95 + 1.795 = 19.745**.

**Float Precision**  
- Python might display the result as **19.745000000000005** due to floating-point precision. That’s normal for floats—some decimals can’t be perfectly represented in binary form.
